### ISSUE TITLE
Unsafe attributes support (`unnecessary_safety_comment`)

### ISSUE THREAD

**Issue by ojeda** [rust-lang/rust-clippy#13317]
### What it does

Unsafe attributes have been stabilized (https://github.com/rust-lang/rust/pull/128771), thus it would be nice to have a lint that ensures `// SAFETY` comments are not written for safe attributes.

Since `unnecessary_safety_comment` covers `impl`s too, perhaps it should also cover this too, although it may be best to possibly split them (but that is a different issue).

### Example

```rust
// SAFETY: ...
#[derive(Debug)]
struct S;
```

Should be written as:

```rust
#[derive(Debug)]
struct S;
```

**Comment by mikeleany** [comment]
It would definitely be nice if this lint supported unsafe attributes, but it seems the lack of support (at least currently) is causing the opposite issue to the one described in the description. The example code in the description triggers the lint in the current stable version, which I would have expected to be the case even before unsafe attributes. What is problematic, though, is that we get a false-positive lint trigger when we include a safety comment on an unsafe attribute like in the following code:

```rust
// SAFETY: shouldn't trigger lint
#[unsafe(no_mangle)]
extern "C" fn not_mangled() {}
```

[Playground link with lint triggering on both safe and unsafe attributes](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e54c15006bb9e98daeb2e0595f216c40)

**Comment by ojeda** [comment]
> The example code in the description triggers the lint in the current stable version, which I would have expected to be the case even before unsafe attributes.

It triggers, but Clippy mentions the `struct`, i.e. Clippy seems to treat the safety comment as applying to the struct rather than the attribute (if one removes the attribute, the message is the same).

> What is problematic, though, is that we get a false-positive lint trigger when we include a safety comment on an unsafe attribute like in the following code

Similarly, in this case, the attribute being there doesn't seem to change anything.

So I think to implement both this issue and https://github.com/rust-lang/rust-clippy/issues/13316 we will need support for tracking safety comments applied to attributes.

**Comment by mikeleany** [comment]
> > The example code in the description triggers the lint in the current stable version, which I would have expected to be the case even before unsafe attributes.
> 
> It triggers, but Clippy mentions the `struct`, i.e. Clippy seems to treat the safety comment as applying to the struct rather than the attribute (if one removes the attribute, the message is the same).
> 
> > What is problematic, though, is that we get a false-positive lint trigger when we include a safety comment on an unsafe attribute like in the following code
> 
> Similarly, in this case, the attribute being there doesn't seem to change anything.
> 
> So I think to implement both this issue and [#13316](https://github.com/rust-lang/rust-clippy/issues/13316) we will need support for tracking safety comments applied to attributes.

Right, the lack of support for safety comments on unsafe attributes is the root of the problem, I was just noting more clearly what the symptoms are: that you get a false positive rather then a false negative.

As for the error message referencing the struct, I don't see that as problematic as I see the safety comment applying to the entire declaration of which the attribute is just one part. However, it might be a good for the error message to include the attribute too since it is part of the declaration. For #13316, we would definitely want the error message to point to the undocumented unsafe attribute.

**Comment by ojeda** [comment]
> As for the error message referencing the struct, I don't see that as problematic as I see the safety comment applying to the entire declaration of which the attribute is just one part. However, it might be a good for the error message to include the attribute too since it is part of the declaration.

Yeah, when I originally thought about this, I was thinking that unsafe attributes would have their own safety comment, i.e. independent of the item, and even that each unsafe attribute would have its own safety comment.
